src/fastalloc/mod.rs

```diff
@@ -1260,6 +1419,67 @@ impl<'a, F: Function> Env<'a, F> {
+            for (idx, pred_succ) in func.block_succs(pred).iter().enumerate() {
+                if *pred_succ == block {
+                    return idx;
+                }
+            }
+            unreachable!(
+                "{:?} was not found in the successor list of its predecessor {:?}",
+                block, pred
+            );
+        };
+        trace!(
+            "Checking for predecessor branch args defined in the branch with fixed-reg constraint"
+        );
+        for (param_idx, block_param) in self.func.block_params(block).iter().cloned().enumerate() {
+            // Block param is never used. Don't bother.
+            if self.vreg_spillslots[block_param.vreg()].is_invalid() {
+                continue;
+            }
+            for pred in self.func.block_preds(block).iter().cloned() {
+                let pred_last_inst = self.func.block_insns(pred).last();
+                let curr_block_succ_idx = get_succ_idx_of_pred(pred, self.func);
+                let branch_arg_for_param =
+                    self.func
+                        .branch_blockparams(pred, pred_last_inst, curr_block_succ_idx)[param_idx];
+                // If the branch arg is defined in the branch instruction, the move will have to be done
+                // here, instead of at the end of the predecessor block.
+                let move_from = self.func.inst_operands(pred_last_inst)
+                    .iter()
+                    .find_map(|op| if op.kind() == OperandKind::Def && op.vreg() == branch_arg_for_param {
+                        match op.constraint() {
+                            OperandConstraint::FixedReg(reg) => {
+                                trace!("Found one for branch arg {branch_arg_for_param} and param {block_param} in reg {reg}");
+                                Some(Allocation::reg(reg))
+                            },
+                            // In these cases, the vreg is defined directly into the block param
+                            // spillslot.
+                            OperandConstraint::Stack | OperandConstraint::Any => None,
+                            constraint => panic!("fastalloc does not support using any-reg or reuse constraints ({}) defined on a branch instruction as a branch arg on the same instruction", constraint),
+                        }
+                    } else {
+                        None
+                    });
+                if let Some(from) = move_from {
+                    let to = Allocation::stack(self.vreg_spillslots[block_param.vreg()]);
+                    trace!("Inserting edit to move from {from} to {to}");
+                    self.add_move(
+                        first_inst,
+                        from,
+                        to,
+                        block_param.class(),
+                        InstPosition::Before,
+                    )?;
+                }
+                break;
+            }
+        }
```

Comment:
Can we check somewhere above that there is only one predecessor in the case that a branch's def forces a move insertion here? That should always be the case when such defs are on conditional branches (because we require split critical edges) and is thus sufficient for Cranelift's `try_call` at least; and it avoids the incorrectness that would otherwise result.